src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -407,5 +406,176 @@ public static void SetLastWriteTime(string fullPath, DateTimeOffset time, bool a
+                switch (error)
+                {
+                    case Interop.Errors.ERROR_FILE_NOT_FOUND:
+                    case Interop.Errors.ERROR_PATH_NOT_FOUND:
+                        if (throwOnNotFound)
+                        {
+                            throw Win32Marshal.GetExceptionForWin32Error(
+                                // File not found doesn't make much sense coming from a directory.
+                                isDirectory ? Interop.Errors.ERROR_PATH_NOT_FOUND : error, linkPath);
+                        }
+                        return null;
+                    default:
+                        throw Win32Marshal.GetExceptionForWin32Error(error, linkPath);
+                }
+            }
+
+            byte[]? buffer = null;
+            try
+            {
+                buffer = ArrayPool<byte>.Shared.Rent(Interop.Kernel32.MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
+                bool success = Interop.Kernel32.DeviceIoControl(
+                    handle,
+                    dwIoControlCode: Interop.Kernel32.FSCTL_GET_REPARSE_POINT,
+                    lpInBuffer: IntPtr.Zero,
+                    nInBufferSize: 0,
+                    lpOutBuffer: buffer,
+                    nOutBufferSize: Interop.Kernel32.MAXIMUM_REPARSE_DATA_BUFFER_SIZE,
+                    out _,
+                    IntPtr.Zero);
+
+                if (!success)
+                {
+                    int error = Marshal.GetLastWin32Error();
+                    // The file or directory is not a reparse point.
+                    if (error == Interop.Errors.ERROR_NOT_A_REPARSE_POINT)
+                    {
+                        return null;
+                    }
+
+                    throw Win32Marshal.GetExceptionForWin32Error(error, linkPath);
+                }
+
+                ReadOnlySpan<byte> bufferSpan = new(buffer);
+                ref readonly Interop.Kernel32.REPARSE_DATA_BUFFER rdb = ref MemoryMarshal.AsRef<Interop.Kernel32.REPARSE_DATA_BUFFER>(bufferSpan);
+
+                // Only symbolic links are supported at the moment.
+                if ((rdb.ReparseTag & Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_SYMLINK) == 0)
+                {
+                    return null;
+                }
+
+                int substituteNameOffset = sizeof(Interop.Kernel32.REPARSE_DATA_BUFFER) + rdb.ReparseBufferSymbolicLink.SubstituteNameOffset;
+                int substituteNameLength = rdb.ReparseBufferSymbolicLink.SubstituteNameLength;
+
+                ReadOnlySpan<char> targetPath = MemoryMarshal.Cast<byte, char>(bufferSpan.Slice(substituteNameOffset, substituteNameLength));
+
+                // Target path is relative, we need to append the link directory.
+                if (normalize && (rdb.ReparseBufferSymbolicLink.Flags & Interop.Kernel32.SYMLINK_FLAG_RELATIVE) != 0)
+                {
+                    return Path.Join(Path.GetDirectoryName(linkPath.AsSpan()), targetPath);
```

Comment:
Could please clarify why do we need to add the link directory?
Should we return ".\targetPath"?

Reply:
The target is always relative to the link's directory. If we don't append it when we return our FileSystemInfo, it will assume that `targetPath` is relative to the current working directory as all the other IO APIs.

Reply:
So if we create a symlink and we use a relative path we get a relative symlink:
```c#
var link = CreateSymbolicLink("c:\temp\linkName", ".\file.txt");
```
but if we resolve target we get a full path `c:\temp\file.txt`?

I guess we want to see `c:\temp\linkName -> .\file.txt` or `c:\temp\linkName -> file.txt` in PowerShell's dir output.

So I'd expect we return  directly `linkPath` here.

/cc @mklement0

Reply:
@iSazonov, I haven't looked at the source code, but my understanding is that if you want the target path _as defined_ (which may be relative), you'll have to use `FileSystemInfo.LinkTarget`, whereas `FileSystemInfo.FullName` for the full path, and I'm _presuming_ that this distinction maps onto the `normalize` flag in this method.

Reply:
> but if we resolve target we get a full path c:\temp\file.txt?

Yes, that's how `link.ResolveLinkTarget().FullPath` would look like. 

> I guess we want to see c:\temp\linkName -> .\file.txt or c:\temp\linkName -> file.txt in PowerShell's dir output.

I think for such case you have to use property `LinkTarget` (in pseudoode: `link.FullPath -> link.LinkTarget`) which returns the unresolved/defined target path. You shouldn't expect ResolveLinkTarget to return a FileSystemInfo with a relative path in its FullPath.